File: packages/core_dart/test/interaction_test.dart
/// Corpus de `interaction.dart` — ESPEJO 1:1 de `tests/interaction.test.ts`
/// (KRO-74). Construye las composiciones vía `fromJson` para validar también el
/// parse. Cubre resolveTapAction / resolveDetailComposition / resolveTargetRecipe
/// / resolveExpandRecipe / isTappable / opensNewScreen.
import 'package:kromia_core/kromia_core.dart';
import 'package:test/test.dart';

ViewComposition base([Map<String, dynamic> over = const {}]) =>
    ViewComposition.fromJson(<String, dynamic>{
      'recipe': 'compact_avatar',
      'action': 'none',
      'slots': <String, dynamic>{},
      ...over,
    });

final fields = <FieldDefLike>[
  const FieldDefLike(key: 'foto', label: 'Foto', type: 'image', behavior: 'avatar'),
  const FieldDefLike(key: 'nombre', label: 'Nombre', type: 'text'),
  const FieldDefLike(key: 'desc', label: 'Desc', type: 'textarea'),
  const FieldDefLike(key: 'web', label: 'Web', type: 'text', behavior: 'url'),
];

void main() {
  group('resolveTapAction', () {
    test('none → TapNone', () {
      expect(resolveTapAction(base(), const {}), isA<TapNone>());
    });

    test('navigate_to_detail → TapNavigate con detailComposition', () {
      final r = resolveTapAction(
        base({'action': 'navigate_to_detail', 'targetRecipe': 'editorial'}),
        const {},
        ResolveTapOptions(fieldDefs: fields),
      );
      expect(r, isA<TapNavigate>());
      expect((r as TapNavigate).detailComposition.recipe, 'editorial');
    });

    test('modal → TapModal', () {
      final r = resolveTapAction(
          base({'action': 'modal'}), const {}, ResolveTapOptions(fieldDefs: fields));
      expect(r, isA<TapModal>());
    });

    test('expand_inline → TapExpand con recipe + slots', () {
      final r = resolveTapAction(
        base({
          'action': 'expand_inline',
          'expand': {
            'recipe': 'accordion_simple',
            'slots': {
              'body': {'fields': ['desc']},
            },
          },
        }),
        const {},
      );
      expect(r, isA<TapExpand>());
      final e = r as TapExpand;
      expect(e.expandRecipe, 'accordion_simple');
      expect(e.expandSlots['body']!.fields, ['desc']);
    });

    test('expand_inline sin expand → TapNone', () {
      expect(resolveTapAction(base({'action': 'expand_inline'}), const {}),
          isA<TapNone>());
    });

    test('external_link → TapExternal con url del item', () {
      final r = resolveTapAction(
        base({'action': 'external_link', 'linkField': 'web'}),
        const {'web': 'https://ejemplo.com'},
      );
      expect((r as TapExternal).url, 'https://ejemplo.com');
    });

    test('external_link field vacío → url vacía', () {
      final r = resolveTapAction(
          base({'action': 'external_link', 'linkField': 'web'}), const {'web': ''});
      expect((r as TapExternal).url, '');
    });

    test('external_link sin linkField → url vacía', () {
      final r =
          resolveTapAction(base({'action': 'external_link'}), const {'web': 'x'});
      expect((r as TapExternal).url, '');
    });
  });

  group('resolveDetailComposition', () {
    test('targetComposition tiene prioridad sobre targetRecipe', () {
      final vc = base({
        'action': 'navigate_to_detail',
        'targetRecipe': 'editorial',
        'targetComposition': {'recipe': 'momento', 'action': 'none'},
      });
      expect(resolveDetailComposition(vc, fields).recipe, 'momento');
    });

    test('targetComposition sin slots → rellena con auto-slots', () {
      final vc = base({
        'action': 'navigate_to_detail',
        'targetComposition': {'recipe': 'hero_protagonico', 'action': 'none'},
      });
      final detail = resolveDetailComposition(vc, fields);
      expect(detail.recipe, 'hero_protagonico');
      expect(detail.slots['avatar']!.fields, contains('foto'));
      expect(detail.slots['title']!.fields, contains('nombre'));
    });

    test('targetComposition preserva el siguiente salto', () {
      final vc = base({
        'action': 'navigate_to_detail',
        'targetComposition': {
          'recipe': 'hero_protagonico',
          'action': 'modal',
          'targetComposition': {'recipe': 'editorial', 'action': 'none'},
        },
      });
      final detail = resolveDetailComposition(vc);
      expect(detail.action, 'modal');
      expect(detail.targetComposition?.recipe, 'editorial');
    });

    test('legacy targetRecipe → esa receta con auto-slots', () {
      final vc =
          base({'action': 'navigate_to_detail', 'targetRecipe': 'editorial'});
      expect(resolveDetailComposition(vc, fields).recipe, 'editorial');
    });

    test('sin target → fallback auto hero_protagonico', () {
      expect(
          resolveDetailComposition(base({'action': 'navigate_to_detail'}), fields)
              .recipe,
          'hero_protagonico');
    });

    test('targetRecipe inválida (list recipe) → fallback auto', () {
      final vc =
          base({'action': 'navigate_to_detail', 'targetRecipe': 'compact_avatar'});
      expect(resolveDetailComposition(vc, fields).recipe, 'hero_protagonico');
    });
  });

  group('resolveTargetRecipe', () {
    test('galería → momento', () {
      expect(
          resolveTargetRecipe(
              [const FieldDefLike(key: 'g', label: 'G', type: 'array<image>')]),
          'momento');
    });
    test('imagen → hero_protagonico', () {
      expect(
          resolveTargetRecipe(
              [const FieldDefLike(key: 'f', label: 'F', type: 'image')]),
          'hero_protagonico');
    });
    test('behavior avatar → hero_protagonico', () {
      expect(
          resolveTargetRecipe([
            const FieldDefLike(key: 'a', label: 'A', type: 'text', behavior: 'avatar')
          ]),
          'hero_protagonico');
    });
    test('sin imagen → editorial', () {
      expect(
          resolveTargetRecipe(
              [const FieldDefLike(key: 'n', label: 'N', type: 'text')]),
          'editorial');
    });
  });

  group('resolveExpandRecipe', () {
    test('field url → accordion_with_actions', () {
      expect(
          resolveExpandRecipe([
            const FieldDefLike(key: 'w', label: 'W', type: 'text', behavior: 'url')
          ]),
          'accordion_with_actions');
    });
    test('sin enlazable → accordion_simple', () {
      expect(
          resolveExpandRecipe(
              [const FieldDefLike(key: 't', label: 'T', type: 'textarea')]),
          'accordion_simple');
    });
  });

  group('isTappable / opensNewScreen', () {
    test('none → no tappable', () {
      expect(isTappable(base()), false);
    });
    test('navigate_to_detail → tappable + opens new screen', () {
      final vc = base({'action': 'navigate_to_detail'});
      expect(isTappable(vc), true);
      expect(opensNewScreen(vc), true);
    });
    test('modal → tappable + opens new screen', () {
      final vc = base({'action': 'modal'});
      expect(isTappable(vc), true);
      expect(opensNewScreen(vc), true);
    });
    test('expand_inline → tappable pero NO opens new screen', () {
      final vc = base({'action': 'expand_inline'});
      expect(isTappable(vc), true);
      expect(opensNewScreen(vc), false);
    });
    test('external_link → tappable pero NO opens new screen', () {
      final vc = base({'action': 'external_link'});
      expect(isTappable(vc), true);
      expect(opensNewScreen(vc), false);
    });
  });
}

File: packages/core/tests/interaction.test.ts
/**
 * KRO-74 — `resolveTapAction`, `resolveDetailComposition`,
 * `resolveTargetRecipe`, `resolveExpandRecipe`, `isTappable`, `opensNewScreen`.
 */
import { describe, it, expect } from 'vitest';
import {
  resolveTapAction,
  resolveDetailComposition,
  resolveTargetRecipe,
  resolveExpandRecipe,
  isTappable,
  opensNewScreen,
} from '../src/interaction';
import type { ViewComposition } from '../src/types';

const base = (over: Partial<ViewComposition> = {}): ViewComposition => ({
  recipe: 'compact_avatar', action: 'none', slots: {}, ...over,
});

const fields = [
  { key: 'foto',   label: 'Foto',   type: 'image' as const, behavior: 'avatar' }, // behavior necesario para auto-slots
  { key: 'nombre', label: 'Nombre', type: 'text'     as const },
  { key: 'desc',   label: 'Desc',   type: 'textarea' as const },
  { key: 'web',    label: 'Web',    type: 'text'     as const, behavior: 'url' },
];

// ─── resolveTapAction ───────────────────────────────────────────────────────

describe('resolveTapAction', () => {
  it('none → { kind: "none" }', () => {
    expect(resolveTapAction(base(), {})).toEqual({ kind: 'none' });
  });

  it('navigate_to_detail → { kind: "navigate", detailComposition }', () => {
    const r = resolveTapAction(
      base({ action: 'navigate_to_detail', targetRecipe: 'editorial' }),
      {},
      { fieldDefs: fields },
    );
    expect(r.kind).toBe('navigate');
    if (r.kind === 'navigate') {
      expect(r.detailComposition.recipe).toBe('editorial');
    }
  });

  it('modal → { kind: "modal", detailComposition }', () => {
    const r = resolveTapAction(base({ action: 'modal' }), {}, { fieldDefs: fields });
    expect(r.kind).toBe('modal');
  });

  it('expand_inline → { kind: "expand", expandRecipe, expandSlots }', () => {
    const r = resolveTapAction(
      base({ action: 'expand_inline', expand: { recipe: 'accordion_simple', slots: { body: { fields: ['desc'] } } } }),
      {},
    );
    if (r.kind !== 'expand') throw new Error('expected expand');
    expect(r.expandRecipe).toBe('accordion_simple');
    expect(r.expandSlots).toEqual({ body: { fields: ['desc'] } });
  });

  it('expand_inline sin composition.expand → none', () => {
    expect(resolveTapAction(base({ action: 'expand_inline' }), {}).kind).toBe('none');
  });

  it('external_link → { kind: "external", url } desde item', () => {
    const r = resolveTapAction(
      base({ action: 'external_link', linkField: 'web' }),
      { web: 'https://ejemplo.com' },
    );
    if (r.kind !== 'external') throw new Error('expected external');
    expect(r.url).toBe('https://ejemplo.com');
  });

  it('external_link con field vacío → url vacía', () => {
    const r = resolveTapAction(base({ action: 'external_link', linkField: 'web' }), { web: '' });
    if (r.kind !== 'external') throw new Error('expected external');
    expect(r.url).toBe('');
  });

  it('external_link sin linkField → url vacía', () => {
    const r = resolveTapAction(base({ action: 'external_link' }), { web: 'x' });
    if (r.kind !== 'external') throw new Error('expected external');
    expect(r.url).toBe('');
  });
});

// ─── resolveDetailComposition ───────────────────────────────────────────────

describe('resolveDetailComposition', () => {
  it('targetComposition (cadena KRO-94) tiene prioridad', () => {
    const vc = base({
      action: 'navigate_to_detail',
      targetRecipe: 'editorial',
      targetComposition: { recipe: 'momento', action: 'none' },
    });
    const detail = resolveDetailComposition(vc, fields);
    expect(detail.recipe).toBe('momento');
  });

  it('targetComposition sin slots → rellena con auto-slots', () => {
    const vc = base({
      action: 'navigate_to_detail',
      targetComposition: { recipe: 'hero_protagonico', action: 'none' },
    });
    const detail = resolveDetailComposition(vc, fields);
    expect(detail.recipe).toBe('hero_protagonico');
    // auto-slots: foto → avatar, nombre → title
    expect(detail.slots.avatar?.fields).toContain('foto');
    expect(detail.slots.title?.fields).toContain('nombre');
  });

  it('KRO-133 — targetComposition con diseño por BLOQUES propaga layout + slotOverrides + accentPosition', () => {
    const layout = { type: 'container', kind: 'grid', columns: 1, rows: 1,
      children: [{ type: 'slot', slot: 'title', place: { colStart: 1, rowStart: 1 } }] } as any;
    const vc = base({
      action: 'navigate_to_detail',
      targetComposition: {
        recipe: 'editorial', action: 'none',
        slots: { title: { fields: ['nombre'] } },
        layout,
        slotOverrides: { disabled: ['subtitle'] },
        accentPosition: 'left',
      } as any,
    });
    const detail = resolveDetailComposition(vc, fields);
    // El lienzo guardado del detalle DEBE llegar al cliente (antes se perdía → preset).
    expect((detail as any).layout).toEqual(layout);
    expect((detail as any).slotOverrides?.disabled).toEqual(['subtitle']);
    expect((detail as any).accentPosition).toBe('left');
  });

  it('targetComposition preserva el siguiente salto (action + targetComposition anidado)', () => {
    const vc = base({
      action: 'navigate_to_detail',
      targetComposition: {
        recipe: 'hero_protagonico', action: 'modal',
        targetComposition: { recipe: 'editorial', action: 'none' },
      },
    });
    const detail = resolveDetailComposition(vc);
    expect(detail.action).toBe('modal');
    expect((detail as any).targetComposition?.recipe).toBe('editorial');
  });

  it('legacy targetRecipe → usa esa receta con auto-slots', () => {
    const vc = base({ action: 'navigate_to_detail', targetRecipe: 'editorial' });
    const detail = resolveDetailComposition(vc, fields);
    expect(detail.recipe).toBe('editorial');
  });

  it('sin ningún target → fallback auto hero_protagonico', () => {
    const detail = resolveDetailComposition(base({ action: 'navigate_to_detail' }), fields);
    expect(detail.recipe).toBe('hero_protagonico');
  });

  it('targetRecipe inválida (list recipe) → fallback auto', () => {
    const vc = base({ action: 'navigate_to_detail', targetRecipe: 'compact_avatar' });
    const detail = resolveDetailComposition(vc, fields);
    expect(detail.recipe).toBe('hero_protagonico'); // auto-fallback
  });
});

// ─── resolveTargetRecipe ────────────────────────────────────────────────────

describe('resolveTargetRecipe', () => {
  it('con galería → momento', () => {
    expect(resolveTargetRecipe([{ key: 'g', label: 'G', type: 'array<image>' }])).toBe('momento');
  });

  it('con imagen → hero_protagonico', () => {
    expect(resolveTargetRecipe([{ key: 'f', label: 'F', type: 'image' }])).toBe('hero_protagonico');
  });

  it('con behavior avatar → hero_protagonico', () => {
    expect(resolveTargetRecipe([{ key: 'a', label: 'A', type: 'text', behavior: 'avatar' }])).toBe('hero_protagonico');
  });

  it('sin imagen → editorial', () => {
    expect(resolveTargetRecipe([{ key: 'n', label: 'N', type: 'text' }])).toBe('editorial');
  });
});

// ─── resolveExpandRecipe ────────────────────────────────────────────────────

describe('resolveExpandRecipe', () => {
  it('con field url → accordion_with_actions', () => {
    expect(resolveExpandRecipe([{ key: 'w', label: 'W', type: 'text', behavior: 'url' }])).toBe('accordion_with_actions');
  });

  it('sin field enlazable → accordion_simple', () => {
    expect(resolveExpandRecipe([{ key: 't', label: 'T', type: 'textarea' }])).toBe('accordion_simple');
  });
});

// ─── isTappable / opensNewScreen ────────────────────────────────────────────

describe('isTappable / opensNewScreen', () => {
  it('action=none → no tappable', () => {
    expect(isTappable(base())).toBe(false);
  });

  it('action=navigate_to_detail → tappable + opens new screen', () => {
    const vc = base({ action: 'navigate_to_detail' });
    expect(isTappable(vc)).toBe(true);
    expect(opensNewScreen(vc)).toBe(true);
  });

  it('action=modal → tappable + opens new screen', () => {
    const vc = base({ action: 'modal' });
    expect(isTappable(vc)).toBe(true);
    expect(opensNewScreen(vc)).toBe(true);
  });

  it('action=expand_inline → tappable pero NO opens new screen', () => {
    const vc = base({ action: 'expand_inline' });
    expect(isTappable(vc)).toBe(true);
    expect(opensNewScreen(vc)).toBe(false);
  });

  it('action=external_link → tappable pero NO opens new screen', () => {
    const vc = base({ action: 'external_link' });
    expect(isTappable(vc)).toBe(true);
    expect(opensNewScreen(vc)).toBe(false);
  });
});
